tiny_dnn/core/framework/tensor.h

```diff
@@ -0,0 +1,281 @@
+      and/or other materials provided with the distribution.
+
+    * Neither the name of tiny-dnn nor the names of its
+      contributors may be used to endorse or promote products derived from
+      this software without specific prior written permission.
+
+    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+    AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+    IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+    DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+    FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+    DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+    SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+    CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+    OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+    OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+#pragma once
+
+#include "tiny_dnn/core/framework/device.fwd.h"
+
+namespace tiny_dnn {
+
+template<typename U = float_t>
+class Tensor {
+ public:
+    /*
+     * Initializes an empty tensor.
+     */
+    Tensor() {}
+
+    /*
+     * Create a tensor of the given dimension.
+     */
+    explicit Tensor(const cnn_size_t d0,
+                    const cnn_size_t d1,
+                    const cnn_size_t d2,
+                    const cnn_size_t d3) {
+        reshape(d0, d1, d2, d3);
+        resize();
+    }
+
+    // Returns the tensor shape
+    std::vector<cnn_size_t> shape() const { return shape_; }
+    
+    // Returns the value of a specified index in the tensor.
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T& at(const cnn_size_t d0,
+          const cnn_size_t d1,
+          const cnn_size_t d2,
+          const cnn_size_t d3) {
+        return *access_data<T>(d0, d1, d2, d3);
+    }
+
+    template<typename T>
+    const T& at(const cnn_size_t d0,
+                const cnn_size_t d1,
+                const cnn_size_t d2,
+                const cnn_size_t d3) const {
+        return *access_data<T>(d0, d1, d2, d3);
+    }
+
+    // Returns the pointer to a specified index in the tensor
+    // Checked version (throw exceptions for out-of-range error)
+    template<typename T>
+    T* ptr(const cnn_size_t d0,
+           const cnn_size_t d1,
+           const cnn_size_t d2,
+           const cnn_size_t d3) {
+        return access_data<T>(d0, d1, d2, d3);
+    }
+    
+    template<typename T>
+    const T* ptr(const cnn_size_t d0,
+                 const cnn_size_t d1,
+                 const cnn_size_t d2,
+                 const cnn_size_t d3) const {
+        return access_data<T>(d0, d1, d2, d3);
+    }
+
+    // zero-overhead version (same performance to raw pointer access.
+    // have an assertion for out-of-range error)
```

Comment:
@edgarriba I guess your are talking about using function/method tempaltes overload with SFINAE rule. I didn't know about `std::enable_if`. Thanks!

I wish future C++ will have compile-time `static_if` feature though. Metaprogramming should be more easier to be written and read.